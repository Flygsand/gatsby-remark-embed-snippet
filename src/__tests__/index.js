jest.mock(`fs`, () => {
  return {
    existsSync: jest.fn(),
    readFileSync: jest.fn(),
  }
})

const fs = require(`fs`)
const Remark = require(`remark`)
const plugin = require(`../index`)

const remark = new Remark()

describe(`gatsby-remark-embed-snippet`, () => {
  beforeEach(() => {
    fs.existsSync.mockReset()
    fs.existsSync.mockReturnValue(true)

    fs.readFileSync.mockReset()
    fs.readFileSync.mockReturnValue(`const foo = "bar";`)
  })

  it(`should error if missing required config options`, () => {
    const markdownAST = remark.parse(`\`embed:hello-world.js\``)

    expect(() => plugin({ markdownAST })).toThrow(
      `Required option "directory" not specified`
    )
  })

  it(`should error if the specified directory does not exist`, () => {
    fs.existsSync.mockReturnValue(false)

    const markdownAST = remark.parse(`\`embed:hello-world.js\``)

    expect(() => plugin({ markdownAST }, { directory: `invalid` })).toThrow(
      `Invalid directory specified "invalid"`
    )
  })

  it(`should error if an invalid file path is specified`, () => {
    fs.existsSync.mockImplementation(path => path !== `examples/hello-world.js`)

    const markdownAST = remark.parse(`\`embed:hello-world.js\``)

    expect(() => plugin({ markdownAST }, { directory: `examples` })).toThrow(
      `Invalid snippet specified; no such file "examples/hello-world.js"`
    )
  })

  it(`should not modify non-embed inlineCode nodes`, () => {
    const markdownAST = remark.parse(`\`console.log("hi")\``)
    const transformed = plugin({ markdownAST }, { directory: `examples` })

    expect(transformed).toMatchSnapshot()
  })

  it(`should convert embed inlineCode nodes to Prism code blocks`, () => {
    const markdownAST = remark.parse(`\`embed:hello-world.js\``)
    const transformed = plugin({ markdownAST }, { directory: `examples` })

    expect(transformed).toMatchSnapshot()
  })

  describe(`Language detection`, () => {
    it(`should set the correct Prism language for CSS files`, () => {
      fs.readFileSync.mockReturnValue(`html { height: 100%; }`)

      const markdownAST = remark.parse(`\`embed:hello-world.css\``)
      const transformed = plugin(
        { markdownAST },
        {
          directory: `examples`,
        }
      )

      expect(transformed).toMatchSnapshot()
    })

    it(`should set the correct Prism language for HTML files`, () => {
      fs.readFileSync.mockReturnValue(`<html></html>`)

      const markdownAST = remark.parse(`\`embed:hello-world.html\``)
      const transformed = plugin(
        { markdownAST },
        {
          directory: `examples`,
        }
      )

      expect(transformed).toMatchSnapshot()
    })

    it(`should set the correct Prism language for JavaScript files`, () => {
      fs.readFileSync.mockReturnValue(`const foo = "bar";`)

      const markdownAST = remark.parse(`\`embed:hello-world.js\``)
      const transformed = plugin(
        { markdownAST },
        {
          directory: `examples`,
        }
      )

      expect(transformed).toMatchSnapshot()
    })

    it(`should set the correct Prism language for Markdown files`, () => {
      fs.readFileSync.mockReturnValue(`# Hi`)

      const markdownAST = remark.parse(`\`embed:hello-world.md\``)
      const transformed = plugin(
        { markdownAST },
        {
          directory: `examples`,
        }
      )

      expect(transformed).toMatchSnapshot()
    })

    it(`should set the correct Prism language for shell scripts`, () => {
      fs.readFileSync.mockReturnValue(`pwd`)

      const markdownAST = remark.parse(`\`embed:hello-world.sh\``)
      const transformed = plugin(
        { markdownAST },
        {
          directory: `examples`,
        }
      )

      expect(transformed).toMatchSnapshot()
    })

    it(`should set the correct Prism language for YAML files`, () => {
      fs.readFileSync.mockReturnValue(`name: Brian Vaughn`)

      const markdownAST = remark.parse(`\`embed:hello-world.yaml\``)
      const transformed = plugin(
        { markdownAST },
        {
          directory: `examples`,
        }
      )

      expect(transformed).toMatchSnapshot()
    })

    it(`should set the default Prism language for unknown file extensions`, () => {
      const markdownAST = remark.parse(`\`embed:hello-world\``)
      const transformed = plugin(
        { markdownAST },
        {
          directory: `examples`,
        }
      )

      expect(transformed).toMatchSnapshot()
    })
  })
})
