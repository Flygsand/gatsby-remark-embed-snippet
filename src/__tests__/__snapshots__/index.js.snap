// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatsby-remark-embed-snippet Language detection should set the correct Prism language for CSS files 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 24,
              "line": 1,
              "offset": 23,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "html",
          "value": "<div class=\\"gatsby-highlight\\">
        <pre class=\\"language-css\\"><code><span class=\\"token selector\\">html</span> <span class=\\"token punctuation\\">{</span> <span class=\\"token property\\">height</span><span class=\\"token punctuation\\">:</span> 100%<span class=\\"token punctuation\\">;</span> <span class=\\"token punctuation\\">}</span></code></pre>
        </div>",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 24,
          "line": 1,
          "offset": 23,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 24,
      "line": 1,
      "offset": 23,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`gatsby-remark-embed-snippet Language detection should set the correct Prism language for HTML files 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 25,
              "line": 1,
              "offset": 24,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "html",
          "value": "<div class=\\"gatsby-highlight\\">
        <pre class=\\"language-html\\"><code><span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;</span>html</span><span class=\\"token punctuation\\">></span></span><span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;/</span>html</span><span class=\\"token punctuation\\">></span></span></code></pre>
        </div>",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 25,
          "line": 1,
          "offset": 24,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 25,
      "line": 1,
      "offset": 24,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`gatsby-remark-embed-snippet Language detection should set the correct Prism language for JavaScript files 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 23,
              "line": 1,
              "offset": 22,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "html",
          "value": "<div class=\\"gatsby-highlight\\">
        <pre class=\\"language-jsx\\"><code><span class=\\"token keyword\\">const</span> foo <span class=\\"token operator\\">=</span> <span class=\\"token string\\">\\"bar\\"</span><span class=\\"token punctuation\\">;</span></code></pre>
        </div>",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 23,
          "line": 1,
          "offset": 22,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 23,
      "line": 1,
      "offset": 22,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`gatsby-remark-embed-snippet Language detection should set the correct Prism language for Markdown files 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 23,
              "line": 1,
              "offset": 22,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "html",
          "value": "<div class=\\"gatsby-highlight\\">
        <pre class=\\"language-markup\\"><code># Hi</code></pre>
        </div>",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 23,
          "line": 1,
          "offset": 22,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 23,
      "line": 1,
      "offset": 22,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`gatsby-remark-embed-snippet Language detection should set the correct Prism language for YAML files 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 25,
              "line": 1,
              "offset": 24,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "html",
          "value": "<div class=\\"gatsby-highlight\\">
        <pre class=\\"language-yaml\\"><code><span class=\\"token key atrule\\">name</span><span class=\\"token punctuation\\">:</span> Brian Vaughn</code></pre>
        </div>",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 25,
          "line": 1,
          "offset": 24,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 25,
      "line": 1,
      "offset": 24,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`gatsby-remark-embed-snippet Language detection should set the correct Prism language for shell scripts 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 23,
              "line": 1,
              "offset": 22,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "html",
          "value": "<div class=\\"gatsby-highlight\\">
        <pre class=\\"language-bash\\"><code><span class=\\"token function\\">pwd</span></code></pre>
        </div>",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 23,
          "line": 1,
          "offset": 22,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 23,
      "line": 1,
      "offset": 22,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`gatsby-remark-embed-snippet Language detection should set the default Prism language for unknown file extensions 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 20,
              "line": 1,
              "offset": 19,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "html",
          "value": "<div class=\\"gatsby-highlight\\">
        <pre class=\\"language-none\\"><code>const foo = \\"bar\\";</code></pre>
        </div>",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 20,
          "line": 1,
          "offset": 19,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 20,
      "line": 1,
      "offset": 19,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`gatsby-remark-embed-snippet should convert embed inlineCode nodes to Prism code blocks 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 23,
              "line": 1,
              "offset": 22,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "html",
          "value": "<div class=\\"gatsby-highlight\\">
        <pre class=\\"language-jsx\\"><code><span class=\\"token keyword\\">const</span> foo <span class=\\"token operator\\">=</span> <span class=\\"token string\\">\\"bar\\"</span><span class=\\"token punctuation\\">;</span></code></pre>
        </div>",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 23,
          "line": 1,
          "offset": 22,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 23,
      "line": 1,
      "offset": 22,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`gatsby-remark-embed-snippet should not modify non-embed inlineCode nodes 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 20,
              "line": 1,
              "offset": 19,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "inlineCode",
          "value": "console.log(\\"hi\\")",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 20,
          "line": 1,
          "offset": 19,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 20,
      "line": 1,
      "offset": 19,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
